# Exploratory Data Analysis of Credit Risk Data

This notebook analyzes a credit risk dataset to identify patterns associated with loan defaults.
The focus is on data quality, feature distributions, and relationships relevant for modeling.


## Setup & Configuration

This section defines imports, paths, and plotting configuration to ensure
consistent and reproducible visualizations throughout the analysis.


The project root is assumed to be one level above the notebook directory.


In [15]:
from pathlib import Path
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

ROOT = Path.cwd().parent
FIG_DIR = ROOT / "reports" / "figures"
FIG_DIR.mkdir(parents=True, exist_ok=True)

plt.rcParams.update({
    "figure.figsize": (10, 5),
    "figure.dpi": 120,
    "savefig.dpi": 220,
    "axes.titlesize": 14,
    "axes.labelsize": 11,
    "xtick.labelsize": 10,
    "ytick.labelsize": 10,
    "axes.grid": False,
})


def savefig(name: str):
    plt.tight_layout()
    plt.savefig(FIG_DIR / f"{name}.png", bbox_inches="tight")
    plt.close()


## 1. Data Loading


This section loads the dataset using a helper function and performs initial sanity checks.


In [16]:

DATA_PATH = ROOT / "data" / "credit_risk.csv"

df = pd.read_csv(DATA_PATH)

print("Shape:", df.shape)
display(df.head(3))
display(df.sample(3, random_state=42))



FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\Metin\\Documents\\data\\credit_risk.csv'

In [17]:
print(DATA_PATH.resolve())


C:\Users\Metin\Documents\data\credit_risk.csv


## 2. Dataset Overview


The dataset contains a mix of numerical and categorical features related to borrower characteristics and loan attributes.


In [18]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32581 entries, 0 to 32580
Data columns (total 12 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   person_age                  32581 non-null  int64  
 1   person_income               32581 non-null  int64  
 2   person_home_ownership       32581 non-null  object 
 3   person_emp_length           31686 non-null  float64
 4   loan_intent                 32581 non-null  object 
 5   loan_grade                  32581 non-null  object 
 6   loan_amnt                   32581 non-null  int64  
 7   loan_int_rate               29465 non-null  float64
 8   loan_status                 32581 non-null  int64  
 9   loan_percent_income         32581 non-null  float64
 10  cb_person_default_on_file   32581 non-null  object 
 11  cb_person_cred_hist_length  32581 non-null  int64  
dtypes: float64(3), int64(5), object(4)
memory usage: 3.0+ MB


## 3. Missing Values Analysis

This section analyzes the proportion of missing values per feature to assess
data quality and guide preprocessing decisions.


In [24]:
missing = df.isna().mean().sort_values(ascending=False) * 100
missing = missing[missing > 0]


In [25]:
fig, ax = plt.subplots()
ax.bar(missing.index, missing.values)
ax.set_title("Missing Values per Feature (%)")
ax.set_ylabel("Percentage")
ax.set_xlabel("Feature with missing values")
ax.tick_params(axis="x", rotation=45)
ax.grid(axis="y", alpha=0.3)

savefig("missing_values_percent")
plt.show()

missing.round(2)



loan_int_rate        9.56
person_emp_length    2.75
dtype: float64

Missing values are present primarily in `loan_int_rate` (~9.6%) and
`person_emp_length` (~2.8%). Given their proportion, these features will require
imputation or careful handling during preprocessing to avoid data leakage or loss
of information.


## 4. Target Variable


In [26]:
df["loan_status"].value_counts(normalize=True)


loan_status
0    0.781836
1    0.218164
Name: proportion, dtype: float64

## 5. Loan Interest Rate vs Loan Status


### Descriptive Statistics


In [27]:
df.drop(columns=["loan_status"]).describe()


person_age  person_income  person_emp_length     loan_amnt  \
count  32581.000000   3.258100e+04       31686.000000  32581.000000   
mean      27.734600   6.607485e+04           4.789686   9589.371106   
std        6.348078   6.198312e+04           4.142630   6322.086646   
min       20.000000   4.000000e+03           0.000000    500.000000   
25%       23.000000   3.850000e+04           2.000000   5000.000000   
50%       26.000000   5.500000e+04           4.000000   8000.000000   
75%       30.000000   7.920000e+04           7.000000  12200.000000   
max      144.000000   6.000000e+06         123.000000  35000.000000   

       loan_int_rate  loan_percent_income  cb_person_cred_hist_length  
count   29465.000000         32581.000000                32581.000000  
mean       11.011695             0.170203                    5.804211  
std         3.240459             0.106782                    4.055001  
min         5.420000             0.000000                    2.000000  
25%         7.900000             0.090000                    3.000000  
50%        10.990000             0.150000                    4.000000  
75%        13.470000             0.230000                    8.000000  
max        23.220000             0.830000                   30.000000

### Loan Interest Rate by Loan Status


In [28]:
df.groupby("loan_status")["loan_int_rate"].mean()


loan_status
0    10.435999
1    13.060207
Name: loan_int_rate, dtype: float64

In [29]:
df.boxplot(column="loan_int_rate", by="loan_status", grid=False)
plt.title("Loan Interest Rate by Loan Status")
plt.suptitle("")
plt.xlabel("Loan Status (0 = No Default, 1 = Default)")
plt.ylabel("Interest Rate (%)")
savefig("loan_int_rate_by_loan_status")
plt.show()



## 6. Correlation Matrix

This section examines correlations between numerical features to identify
strong relationships and potential multicollinearity.


In [13]:
# Select numerical features only
num_df = df.select_dtypes(include=["int64", "float64"]).copy()

# Compute correlation matrix
corr = num_df.corr()

# Plot
fig, ax = plt.subplots(figsize=(10, 8))
im = ax.imshow(corr.values)

ax.set_title("Correlation Matrix (Numerical Features)")

ax.set_xticks(range(len(corr.columns)))
ax.set_yticks(range(len(corr.columns)))
ax.set_xticklabels(corr.columns, rotation=90)
ax.set_yticklabels(corr.columns)

# Colorbar
fig.colorbar(im, ax=ax, fraction=0.046, pad=0.04)

savefig("correlation_matrix")
plt.show()


In [38]:
print(FIG_DIR)


C:\Users\Metin\Documents\reports\figures


In [39]:
print("ROOT:", ROOT.resolve())
print("FIG_DIR:", FIG_DIR.resolve())


ROOT: C:\Users\Metin\Documents
FIG_DIR: C:\Users\Metin\Documents\reports\figures


## 5. Key Findings


- The dataset contains 32,581 observations with 12 features and no missing values.
- The target variable `loan_status` is moderately imbalanced (~78% non-default, ~22% default).
- Loans that default have higher average interest rates, consistent with risk-based pricing.
- Several numerical features contain extreme outliers and should be handled before modeling.
